include/aws/mqtt/private/request-response/subscription_manager.h

```diff
@@ -0,0 +1,206 @@
+#ifndef AWS_MQTT_PRIVATE_REQUEST_RESPONSE_SUBSCRIPTION_MANAGER_H
+#define AWS_MQTT_PRIVATE_REQUEST_RESPONSE_SUBSCRIPTION_MANAGER_H
+
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/mqtt/mqtt.h>
+
+#include <aws/common/hash_table.h>
+
+struct aws_mqtt_protocol_adapter;
+struct aws_protocol_adapter_connection_event;
+struct aws_protocol_adapter_subscription_event;
+
+/*
+ * The kind of subscription event being emitted.
+ */
+enum aws_rr_subscription_event_type {
+
+    /*
+     * A subscribe succeeded
+     */
+    ARRSET_SUBSCRIPTION_SUBSCRIBE_SUCCESS,
+
+    /*
+     * A subscribe failed
+     */
+    ARRSET_SUBSCRIPTION_SUBSCRIBE_FAILURE,
+
+    /*
+     * A previously successful subscription has ended (generally due to a failure to resume a session)
+     */
+    ARRSET_SUBSCRIPTION_ENDED
+};
+
+struct aws_rr_subscription_status_event {
+    enum aws_rr_subscription_event_type type;
+    struct aws_byte_cursor topic_filter;
+    uint64_t operation_id;
+};
+
+/*
+ * Invariant: despite being on the same thread, these callbacks must be queued as cross-thread tasks on the native
+ * request-response client.  This allows us to iterate internal collections without worrying about external
+ * callers disrupting things by invoking APIs back on us.
+ */
+typedef void(
+    aws_rr_subscription_status_event_callback_fn)(const struct aws_rr_subscription_status_event *event, void *userdata);
+
+struct aws_rr_subscription_manager_options {
+
+    /*
+     * Maximum number of concurrent subscriptions allowed
+     */
+    size_t max_subscriptions;
+
+    /*
+     * Ack timeout to use for all subscribe and unsubscribe operations
+     */
+    uint32_t operation_timeout_seconds;
```

Comment:
No, at present there's a single timeout for all operations.  We added per-operation because we can't control if the mqtt client even has a timeout enabled, let alone what it is.